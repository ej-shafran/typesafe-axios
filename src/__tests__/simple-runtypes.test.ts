import { taxios } from "..";

import * as st from "simple-runtypes";
import nock from "nock";
import * as UTILS from "./__utils__";

const SCHEMA = st.record({
  id: st.string(),
  username: st.string(),
  email: st.string(),
  address: st.record({
    street: st.string(),
    zipcode: st.optional(st.string()),
  }),
  followers: st.array(
    st.record({
      id: st.string(),
      role: st.optional(st.stringLiteralUnion("admin", "guest")),
    })
  ),
});

type TYPE = ReturnType<typeof SCHEMA>;
function ensure(_: TYPE) {}

describe("simple-runtypes interop", () => {
  it("passes along data with the correct type", async () => {
    nock(UTILS.API).get("/").reply(200, UTILS.VALID);

    const { data } = await taxios(SCHEMA).get(UTILS.API);

    expect(data).toEqual(UTILS.VALID);
    ensure(data);
  });

  it("throws a validation error when the data is invalid", async () => {
    nock(UTILS.API).get("/").reply(200, UTILS.INVALID);

    try {
      await taxios(SCHEMA).get(UTILS.API);
      fail("Did not throw error.");
    } catch (error) {
      expect(JSON.stringify(error, null, 2)).toMatchSnapshot();
    }
  });
});
