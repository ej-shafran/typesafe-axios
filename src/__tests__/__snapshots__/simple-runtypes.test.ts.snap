// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple-runtypes interop throws a validation error when the data is invalid 1`] = `
"{
  "name": "RuntypeError",
  "reason": "missing key in record: \\"email\\"",
  "path": []
}"
`;
